routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,185 @@
+#include "routing/bicycle_directions.hpp"
+#include "routing/car_model.hpp"
+#include "routing/router_delegate.hpp"
+#include "routing/routing_result_graph.hpp"
+#include "routing/turns_generator.hpp"
+
+#include "geometry/point2d.hpp"
+
+#include "indexer/ftypes_matcher.hpp"
+#include "indexer/index.hpp"
+#include "indexer/scales.hpp"
+
+namespace
+{
+using namespace routing;
+using namespace routing::turns;
+
+class AStarRoutingResultGraph : public IRoutingResultGraph
+{
+public:
+  AStarRoutingResultGraph(IRoadGraph::TEdgeVector const & routeEdges,
+                          AdjacentEdgesMap const & adjacentEdges,
+                          TUnpackedPathSegments const & pathSegments)
+    : m_routeEdges(routeEdges), m_adjacentEdges(adjacentEdges), m_pathSegments(pathSegments),
+      m_routeLength(0)
+  {
+    for (auto const & edge : routeEdges)
+    {
+      m_routeLength += MercatorBounds::DistanceOnEarth(edge.GetStartJunction().GetPoint(),
+                                                       edge.GetEndJunction().GetPoint());
+    }
+  }
+
+  ~AStarRoutingResultGraph() {}
+
+  virtual TUnpackedPathSegments const & GetSegments() const override
+  {
+    return m_pathSegments;
+  }
+
+  virtual void GetPossibleTurns(TNodeId node, m2::PointD const & ingoingPoint,
+                                m2::PointD const & junctionPoint,
+                                size_t & ingoingCount,
+                                TTurnCandidates & outgoingTurns) const override
+  {
+    if (node >= m_routeEdges.size() || node >= m_adjacentEdges.size())
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    AdjacentEdgesMap::const_iterator adjacentEdges = m_adjacentEdges.find(node);
+    if (adjacentEdges == m_adjacentEdges.cend())
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    ingoingCount = adjacentEdges->second.ingoingTurnCount;
+    outgoingTurns = adjacentEdges->second.outgoingTurns;
+  }
+
+  virtual double GetPathLength() const override { return m_routeLength; }
+
+  virtual m2::PointD const & GetStartPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.front().GetStartJunction().GetPoint();
+  }
+
+  virtual m2::PointD const & GetEndPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.back().GetEndJunction().GetPoint();
+  }
+
+private:
+   IRoadGraph::TEdgeVector const & m_routeEdges;
+   AdjacentEdgesMap const & m_adjacentEdges;
+   TUnpackedPathSegments const & m_pathSegments;
+   double m_routeLength;
+};
+
+ftypes::HighwayClass GetHighwayClass(FeatureID const & featureId, Index const & index)
+{
+  ftypes::HighwayClass highWayClass = ftypes::HighwayClass::Undefined;
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  uint32_t const featureIndex = featureId.m_index;
+
+  FeatureType ft;
+  Index::FeaturesLoaderGuard loader(index, mwmId);
```

Comment:
>  I see here that FeaturesLoaderGuard is used for bicycle turns generation, car routing is irrelevant here.

For turn generation of car routing we need to get feature properties of routes as well. So we do use FeaturesLoaderGuard for every possible bifurcation point.

> FeaturesLoaderGuard requires at least one file open/close operation + mutex lock unlock. It's not cheap, right?
> In any case, please, optimize FeatureLoaderGuard usage or get rid of this function at all.

Ok. We'll do that.

> ... we're fanatically and without any reason require std::unordered_map instead of std::maps. 

Are you sure that it's a good idea?